build: Depends qt6 and cmake compatibility

jamescowens opened the pull request:
This PR updates critical components of the depends environment to Bitcoin Core 30.0, including the updating of depends Qt to Qt6 (currently at 6.7.3, the same as Bitcoin Core). Note that it also makes the necessary changes to the depends system to emit toolchain.cmake, and overhauls the cmake build system to be completely ready for the pending removal of the autotools build system. Currently Qt5 is still supported in a local build (for local use or packaging) via cmake on Linux, but this will be removed in the future once Qt5 is sunsetted.

Key Changes:
- **Qt6 Compatibility** Upgrades the depends system to use Qt6 and
  align closer with the current depends in Bitcoin Core v30.0.
- **Build System Unification:** The CMake configuration now supports 
  Linux Native, Linux Static (depends), Windows Cross-Compile, and macOS Native
  targets with a single consistent workflow.
- **Test Framework Fixes:** Resolved linking issues with Boost Test 
  by conditionally applying dynamic/static linkage macros based on 
  the target. Tests now build and pass on all platforms.
- **Static Linking:** Refined `src/test/CMakeLists.txt` and dependencies 
  to correctly handle transitive dependencies for static builds.
- **Depends:** Updated `boost.mk` to ensure the Unit Test Framework 
  library is compiled and installed correctly.
- **Overhaul of CI/CD ** Complete overhaul of CI/CD, integrating CD into CI as envisioned by #2689.
- ** Implementation of several helper scripts:**
  - build_targets.sh: A relatively friendly script that helps users do local compiles for three major targets, Linux Native, Linux Depends, and Windows Cross Compile (depends).
  - install_dependencies.sh: A helper script that works with build_targets.sh to assist in checking/installing dependencies for the major distributions: Debian, Ubuntu, Mint, Fedora/RHEL, OpenSUSE/SLES, and Arch/Manjaro. Note that Ubuntu, Fedora, and OpenSUSE (Leap and Tumbleweed) have been pretty thoroughly tested already.
  - contrib/devtools/run-local-ci.sh: A wrapper script around ACT primarily for executing CI jobs locally in docker containers, with concurrency limits. (Does NOT support native macOS jobs, obviously.)
  - contrib/devtools/split-logs.py: A short python program to split a combined output log from a run-local-ci.sh where several jobs are run in parallel.

###  CI/CD & Build System Updates (Summary)

As part of the CMake migration, the CI pipelines have been overhauled to be **hermetic, reproducible, and developer-friendly**.

**1. Clear separation of purpose with CI configuration files**
* There are now four separate Github Actions configuration files:
   - **cmake_quality.yml** for quality control. There are two jobs: Sanitizers and the Linter.
   - **cmake_production.yml** for integrated CI and CD jobs that will produce artifacts and publish to releases. These currently consist of three jobs: Linux Static (Depends), Windows Cross Compile, and macOS Native (Homebrew Qt6). Each of these produces an artifact that will be downloadable on each CI invocation and published to tags (releases).
   - **cmake_compatibility.yml** for architecture compatibility checking. These currently consist of three jobs: Linux Native (System Qt6), Linux ARM64 (System Qt6), and Linux ARMHF (System Qt6). This checks compatibility with Qt6 and also uncovers architectural and word-size problems (i.e. Intel vs ARM and 64 vs 32 bit).
   - **cmake_distros.yml** for distro compatibility checking. These currently consist of six jobs: Latest Fedora (42), OpenSUSE Lep 15.6, OpenSUSE Tumbleweed, Arch Linux, Debian Unstable (Sid), and Linux Mint. This covers all three major packaging schemes, rpm, apt, and Packman, and covers some bleeding edge versus conservative distros. This serves as validation of the user build_targets.sh and install_dependencies.sh, since these scripts are used in this CI job set.
* All jobs other than the Sanitizer and Linter are full builds with GUI. Qt6 is used on all CI jobs. The test suite is run on both core and GUI components. This includes ARM, where the tests are run via QEMU emulation.

**2. Hermetic Windows Cross-Compilation**
* **Problem:** The previous build relied on implicit system paths to find Qt host tools (`moc`, `rcc`), causing failures when the runner environment didn't match the `depends` system exactly.
* **Fix:** We now explicitly set `QT_HOST_PATH` and `QT_HOST_PATH_CMAKE_DIR` in the CMake configuration.
* **Result:** The build now strictly uses the tools built in `depends/<host>/native`, completely ignoring the host OS environment. This guarantees consistency between CI and local builds.

**3. Local Execution Support (`act`)**
* **Problem:** Running the full CI matrix locally (via `act`) spawned ~72 concurrent compilation threads, causing swap exhaustion and UI freezes on developer workstations.
* **Fix:** Implemented dynamic concurrency control.
    * Added a `Setup Concurrency` step to all workflows.
    * Workflows now respect `CMAKE_BUILD_PARALLEL_LEVEL` if injected externally.
    * Added `run-local-ci.sh` wrapper that auto-scales threads (e.g., `nproc / # of jobs`) when running multiple jobs in parallel.

**4. Test Stability (Wine/Windows)**
* **Problem:** Windows unit tests were crashing or hanging in the Linux CI environment.
* **Fix:**
    * Enforced `WINEARCH=win64` and installed `wine64` packages to fix "Bad EXE format" errors.
    * Forced `QT_QPA_PLATFORM=offscreen` to prevent crashes when tests attempt to create GUI windows on headless runners.
    * Pre-initialized `wineboot` to prevent timeouts during test execution.

**5. Packaging Fixes**
* **Fix:** Updated `src/qt/CMakeLists.txt` to use absolute paths (`${CMAKE_CURRENT_SOURCE_DIR}`) for icons and desktop files. This resolves CPack errors on macOS and Linux where relative paths were resolving incorrectly during the `install` phase.

**5. Continuous Deployment**
* **Feature:** Added a `deploy` job that triggers only on version tags (`v*`). It aggregates artifacts from Linux/Windows/macOS builds and publishes a Draft Release with checksums.

### Documentation:
- **New Guide:** Added `doc/build.md` as the canonical build guide.
- **Migration:** Added a detailed Autotools-to-CMake migration table 
  explaining build types (RelWithDebInfo vs Release) and flags.
- **Platform Docs:** - Updated `doc/build-windows.md` to feature WSL cross-compilation.
  - Deprecated `doc/build-msys2.md` and `doc/build-unix.md`.
  - Added experimental CMake instructions for macOS and BSDs.

The removal of autotools style build capability will be in in a separate PR.

~~@CyberTailor @div72 if you want to help out on this, please put up PR's against my source branch for this PR on my repo clone, and when I accept them they will flow to this one. In particular @div72, we need to adjust CI to add the windows builds to the cmake CI side and remove the depends builds from the non-cmake CI side.~~

Update: I have now done the core of the CI revisions. Please see the comments on the three new files below.

~~The complete removal of the non-cmake CI jobs should be reserved for the separate PR for the retirement of the autotools build for Gridcoin.~~

The old CI files have been deactivated and new CD has been integrated into CI.

~~Also, we need to get the Windows installer packaging working.~~

Integration of the existing share/setup.nsis.in script has been done with a custom cmake target "deploy".

jamescowens committed:
Update critical portions of depends to Bitcoin Core 30.0

This commit also updates depends to Qt6, with all of the changes
necessary. Depends now emits the toolchain.cmake file, which
is used to do a Windows cmake cross compile build.

Minor changes to Gridcoin's CMakeLists.txt files and a few
adjustments to source code in a couple of places have been
done.

jamescowens committed:
Check for stale src/gridcoin-config.h

This adds a block in src/CMakeLists.txt to check for a stale
leftover src/gridcoin-config.h from an autotools in-source build
and delete, since it will cause any number of build headaches.

It also cleans up a few stray working comments.

jamescowens committed:
Configure SVG capabilities on for Qt6 (for icons)

jamescowens committed:
Add block in src/qt/CMakeLists.txt to clear stale in-tree MOC files

CyberTailor commented:
Thanks! I have also made an attempt to make the depends system work with cmake (https://github.com/gridcoin-community/Gridcoin-Research/pull/2812#issuecomment-3055618692) but never finished it. Some results can be seen here: https://github.com/CyberTailor/Gridcoin-Research/tree/my-depends-cmake 

> Also, we need to get the Windows installer packaging working.

This seems to be easy: `setup.nsi.in` becomes `NSIS.template.in` and gets used by CPack. I've edited the template accordingly but haven't tested it yet.

When I have free time, I'll look at this.

jamescowens commented:
I have got Windows building correctly now and also checked native cmake compile for regression and it works. Going to work on Linux depends next. You and Div72 can handle the NSIS part and also the realignment of CI.

jamescowens committed:
Change depends Makefile c standard back to gnu11

jamescowens committed:
Disable unneeded http2 in depends CURL build

jamescowens committed:
Add and/or modify packages for hermetic Linux depends build

jamescowens committed:
Add libxcb_util_cursor.mk required for Qt6 and make other adjustments

jamescowens committed:
Add libXdmcp.mk

jamescowens committed:
Update qt.mk to look in the right place for X related packages

jamescowens committed:
Make adjustements to qt.mk and libXdmcp.mk to fix missing deps and linkage parameters

jamescowens committed:
Add bzip2 and xz to depends

jamescowens committed:
Remove unnecessary functionality from depends curl build

jamescowens committed:
Adjustments to support linux depends build

jamescowens committed:
Fix cmake build special instruction detection (SSE, AVX, SHANI)

This fixes the cmake special optimizations for CPU acceleration for native
compilation.

jamescowens committed:
Enable special acceleration instruction checks for cross-compilation

jamescowens committed:
Correction to conditional for static library linkage in depends build for Linux

jamescowens committed:
Fix Windows depends build regressions after the changes to get Linux depends build to work

jamescowens committed:
Change help text to Gridcoin from Bitcoin Core in depends Makefile

jamescowens committed:
Revent conditional change in C standard in depends Makefile

jamescowens committed:
Tweak to qt.mk

jamescowens commented:
As of https://github.com/gridcoin-community/Gridcoin-Research/pull/2826/commits/e31f7bba5a8500cf788a819d9ef564f04d7b2a38 I have all three major targets working correctly with cmake.

Linux native build with cmake.
Linux depends build with cmake.
Windows depends build with cmake.

Also, here are the Gridcoin startup timings to show that the SSE/SHANI detection is working properly. Note the CMAKE_BUILD_TYPE flags must be used to get autotools default like performance, because cmake by default uses NO optimization, whereas autotools uses -02.

[build_type_timings.xlsx](https://github.com/user-attachments/files/23684856/build_type_timings.xlsx)

jamescowens committed:
Changes to get Gridcoin tests working in both native and depends builds

jamescowens committed:
Initial build documentation revisions for move to cmake

jamescowens committed:
Integrate Windows NSIS installer into cmake build system for Gridcoin

jamescowens committed:
Fixes for Windows and Linux depends build test suite and Windows resources

This fixes the linking to ensure the Gridcoin test suite is
build properly in the cmake cross-compile build for Windows
and the Linux depends build.

Note that DCMAKE_CROSSCOMPILING_EMULATOR=/usr/bin/wine must be
used to allow invocation of the test suite with

ctest --test-dir build_win64

jamescowens committed:
Minor changes to build docs based on PR testing

jamescowens committed:
Prevent qt config in depends from looking for LibWrapClang

Gridcoin does not need this Clang library for language translations,
and upgraded OS's may leave stray instances of Clang contig files that
will trip up compilation otherwise.

jamescowens committed:
Don't hardcode location of libm.so.6 for linux depends builds

jamescowens committed:
Add build-dependencies.md and update build.md and cmake-options.md

jamescowens committed:
Proper support for Boost 1.89+ on new distros

Changes to root CMakeLists.txt to ensure new distros with only
Boost 1.89 do not fail on check for boost system header.

jamescowens committed:
Add -DCMAKE_DISABLE_FIND_PACKAGE_ZSTD=ON to boost.mk to prevent dependency leakage to host

jamescowens committed:
Change zlib to cmake and pass compiler flags to pass with MinGW GCC 14+ compilers

jamescowens committed:
Add build_target.sh and install_dependencies.sh helper scripts

These two scripts simplify the build process for the three main targets
for Linux distributions. Note that the install_dependencies.sh
script is a work in progress.

jamescowens committed:
Modify depends zlib.mk to ensure shared libraries are not built

The cmake build for zlib is ignoring the -DBUILD_SHARED_LIBS=OFF
cmake parameter and building them anyway. This causes problems
with the main Gridcoin build, because the cmake linker there will
prefer the shared library if it exists. Here we manually delete
the generated shared libraries for zlib and ensure the static
library has the expected name. This is hacky but necessary to
ensure we have a hermetically sealed Linux depends binary.

jamescowens committed:
Deactivate legacy CI

jamescowens committed:
Add new cmake production CI targets

This introducesa new CI yaml file and also modifies
install_dependencies.sh to ensure all dependencies are met.

jamescowens committed:
Changes to install_dependencies.sh (WIP)

jamescowens committed:
Change to qt.mk for CI build

jamescowens committed:
More tweaks to fix CI depends builds

jamescowens committed:
More tweaks (2)

jamescowens committed:
Force posix threading on Ubuntu for Windows cross-compile

jamescowens committed:
Add patch package for Fedora

jamescowens committed:
Add package perl-FindBin to Fedora

jamescowens committed:
Add packages bison and flex for Fedora

jamescowens committed:
Add libstdc++-static package for Fedora

jamescowens committed:
Define DEP_LIB for CI linux depends build

jamescowens committed:
Add package xxd to Fedora to support running test suite in Windows cross-compile (XXD)

jamescowens committed:
Initial try at adding MacOS to cmake_production.yml CI jobs

jamescowens committed:
Disable cmake-ci

jamescowens committed:
Add -DHAVE_IOCTLSOCKET -DHAVE_IOCTLSOCKET_FIONBIO to fix curl compile problem on Fedora

jamescowens committed:
corrections to truncated macOS CI job (TBS)

jamescowens committed:
Initial replacement for old CI sanitizer quality checks

jamescowens committed:
Add export LC_ALL=C to new scripts to satisfy linter

jamescowens committed:
Minor corrections for linter

jamescowens committed:
Do not use assembly in Sanitizer CI job

jamescowens commented:
I have implemented new CI jobs.

jamescowens committed:
Introduce cmake_compatibility.yml CI file

Move native Linux build to this file from cmake_production.yml.

jamescowens committed:
Add arm Gridcoin tests in aarch64 build

CI: Add Linux ARM64 Qt6 cross-compilation workflow with QEMU testing

This commit introduces a new GitHub Actions workflow (cmake_compatibility.yaml)
focused on cross-compiling Gridcoin for Linux ARM64 using CMake and Qt6.

Key features and fixes:
- **Cross-Compilation:** Builds for aarch64-linux-gnu on an ubuntu-24.04 runner.
- **Qt6 GUI Support:** Enables GUI build (-DENABLE_GUI=ON -DUSE_QT6=ON) and
  installs required Qt6 modules (Charts, SVG, Core5Compat).
- **Dependency Strategy:** Adopts a "Pure Foreign" installation strategy.
  conflicting native (AMD64) libraries (libcups2t64, libcurl4) are purged
  before installing ARM64 development headers to bypass Ubuntu Multiarch
  version locking issues.
- **Native Tools:** Side-loads native `libzip-tools` and `qemu-user-static`
  to satisfy CMake build requirements and emulation needs.
- **Emulated Testing:** Configures CTest to use `qemu-aarch64-static`
  automatically.
- **Headless GUI Tests:** Sets `QT_QPA_PLATFORM=offscreen` to allow
  `gridcoin_qt_tests` to pass in a headless CI environment.
- **Optimization:** Uses `$(nproc)` for parallel builds and testing to
  maximize runner utilization.

Verified locally via `act` with 100% pass rate on both core and Qt tests.

Also changed test invocation to Use parallelism in test suite runs

jamescowens committed:
Provide run-local-ci.sh developer script to allow local runs of CI jobs with ACT

jamescowens committed:
Cosmetic fixes for cmake CI job friendly names

jamescowens commented:
There are now three CI files:

cmake_production.yml for "production" runs which correspond to build artifacts produced in CD
cmake_compatibility.yml for dynamically linked runs to native x86_64 runner os (currently Ubuntu 24.04) or arm64 cross-compile
cmake_quality.yml for the sanitizers run and the linter

Note that the arm64 cross compile, Linux ARM64 (System Qt6) is a full cross-compile build, including the daemon, the GUI, and both core and qt tests. The tests are run in the QEMU ARM64 emulator.

jamescowens committed:
Add Linux ARMhf (ARM 32 bit) CI build and test

jamescowens commented:
I just did the armhf (ARM 32 bit) CI job. It is a full blown cross-build and test with the arm emulator. This is important to retain for a few reasons:

1. We need an actual test to ensure that 64-bit brittleness does not creep into our code, which can easily happen if you are not actively building and testing on 32-bit targets. Some mistakes can actually affect consensus. Having the 32 bit ARM builds in the old CI was one of the reasons our code is so clean in this respect.

2. Some people are still running 32 bit ARM devices, typically as enthusiast SoC's such as Odroids.

I do NOT intend to create a 32 bit Intel CI job as this is a much narrower case utility-wise, and it is a much more limited test than the ARM 32-bit CI job.

jamescowens committed:
Implementation of CD integrated into new CI.

This is a reimplementation of #2689 contextualized
to this PR.

jamescowens committed:
Augment run-local-ci.sh to handle matrix subjobs.

jamescowens committed:
Fix icon resource path issues

jamescowens committed:
Harden production CI pipeline for local ACT execution and CD

- CI: Enable release deployment on tag push (v* or 5.*)
- CI: Skip artifact uploads when running locally via ACT
- Windows: Fix cross-compilation by forcing usage of depends-built Qt6 host tools (QT_HOST_PATH)
- Windows: Fix test instability by enforcing WINEARCH=win64 and installing wine64
- macOS: Fix packaging path logic in cpack step

jamescowens committed:
Implement dynamic concurrency control for local execution of CI with ACT

This commit standardizes the build parallelism logic across all CMake workflows to
support resource-constrained local execution via 'act'.

Key Changes:
- Added a "Setup Concurrency" step to all jobs (Production, Compatibility, Quality).
- This step defaults to '$(nproc)' (all cores) for standard GitHub Runners but
  respects 'CMAKE_BUILD_PARALLEL_LEVEL' if injected externally.
- Removed hardcoded '--parallel $(nproc)' flags from CMake and CTest commands.
- Allows local 'act' runners to cap threads (e.g., nproc/number of jobs) to prevent
  swap exhaustion when running multiple matrix jobs simultaneously.

jamescowens committed:
Fix run-local-ci.sh to properly calculate per job parallelism in matrix filter case

jamescowens committed:
Change export LC_ALL=C to export LC_ALL=C.UTF-8 in build_targets.sh

This is necessary to prevent compilation errors for Linux depends
on some distributions.

jamescowens committed:
Remove old cmake ci config files

jamescowens committed:
Revise CD artifact preparation

jamescowens committed:
Revise SHA256 checksum generation for CD

jamescowens committed:
Flatten artifacts structure for CD

jamescowens committed:
Add hermeticity check for Linux depends binaries

The binaries produced by the Linux depends build are hybrid static
binaries. They use only the most fundamental libraries on a Linux
system, namely

linux-vdso.so.1 (Kernel interface)
libm.so.6 (Math)
libc.so.6 (C Runtime)
ld-linux-x86-64.so.2 (Dynamic Loader)

This added check verifies that the binaries built ONLY link to these
and no others. This will protect against "leakage" in the future
where something is added which causes a reference to a system
dynamic library to be linked, by causing the CI to fail, which will
force inspection and correction of the leakage.

jamescowens committed:
Modify name of cmake production builds

jamescowens committed:
ci: refine local concurrency heuristic in run-local-ci.sh

This commit improves the matrix detection logic in the local CI wrapper to prevent
over-estimation of job counts, and also fixes concurrency problems when running
locally by ACT.

Key Changes:
- Updated the regex heuristic to strictly match matrix keys ('distro', 'host')
  while excluding generic keys like 'name' or 'image'.
- This fixes a bug where workflow steps named "name:" were falsely counted as
  parallel matrix jobs, causing the script to calculate an incorrect CPU
  allocation (e.g., dividing by 7 instead of 6).
- Also fixes a bug where the staging directory is subject to a race by different
  parallel jobs.

jamescowens committed:
ci: add multi-distro validation workflow and dependency fixes

This commit introduces a comprehensive validation pipeline to verify the
'install_dependencies.sh' and 'build_targets.sh' scripts across major Linux
distribution families.

Key Changes:
1. New Workflow (.github/workflows/cmake_distros.yml):
   - Runs full native builds inside Docker containers for Fedora, OpenSUSE Leap,
     OpenSUSE Tumbleweed, Arch Linux, Debian Sid, and Linux Mint.
   - Validates that the dependency installation script functions correctly on
     bare-metal/minimal environments.

2. Dependency Script Fixes (install_dependencies.sh):
   - OpenSUSE: Added granular Boost development packages (filesystem, thread, etc.).
   - OpenSUSE: Added conditional logic for Tumbleweed (where Boost System is
     header-only) vs Leap (where libboost_system-devel is required).
   - OpenSUSE Leap: Added 'gcc13' installation to support C++17 requirements.
   - Arch Linux: Enforced full system upgrade ('pacman -Syu') to prevent shared
     library mismatches (specifically ICU versions).
   - Corrected package names for Qt6 compatibility modules.

jamescowens committed:
ci: harden local execution, production release, and distro support

This commit finalizes the CI/CD infrastructure by adding multi-distribution
validation, fixing local execution isolation issues, and hardening the
release artifact generation.

Key Changes:

1. Local Execution Wrapper (run-local-ci.sh):
   - Implemented strict isolation using 'rsync' with '--chmod=u+w' to prevent
     read-only Git object permission errors during copying. The purpose of the
     isolation is to avoid polluting the source tree when running ci locally
     via containers with act.
   - Fixed rsync exclusion logic to explicitly include 'build_targets.sh' while
     excluding build directories.
   - Added 'HOST_WORKSPACE' environment variable injection to support correct
     volume mounting when running Docker-in-Docker via 'act'.
   - Added 'act_jobs' parameter to override default container concurrency.
   - Added ccache host mapping for performance.

2. Distro Validation (cmake_distros.yml):
   - Added new workflow to validate builds on Fedora, Arch, OpenSUSE (Leap/Tumbleweed),
     Debian Sid, and Linux Mint.
   - Implemented "Read-Only Mount + Internal Copy" strategy using 'HOST_WORKSPACE'
     to prevent host source tree pollution and race conditions during parallel runs.

3. Dependency Management (install_dependencies.sh):
   - Ubuntu/Debian: Added 'xxd' (required for Windows cross-compile).
   - OpenSUSE Leap: Added 'libdb_cxx-4_8-devel' and forced 'gcc-13' for C++17 support.
   - OpenSUSE Tumbleweed: Added conditional logic to handle header-only Boost System.
   - Arch Linux: Enforced full system upgrade ('-Syu') and added 'icu' to fix link errors.
   - Fixed 'qt6-linguist-devel' package name for OpenSUSE.

4. Production & Release (cmake_production.yml):
   - Hardened 'deploy' job to recursively flatten artifacts and clean up empty
     directories before checksumming, fixing "Is a directory" failures.
   - Added step to copy/rename the Windows GUI executable to the release assets.
   - Added hermeticity check (ldd) for Linux artifacts.

5. Compatibility Workflow (cmake_compatibility.yml):
   - Assigned unique build directories (build_native, build_arm64, build_armhf)
     to prevent collisions when running parallel jobs locally on shared volumes.
   - Enabled ccache.

6. Build Script (build_targets.sh):
   - Added 'PARALLEL' argument for explicit thread control.
   - Added ccache support for the 'depends' build phase.

jamescowens committed:
Remove old CI/CD files

jamescowens committed:
Add split-logs.py utility

This utility separates a single intermixed log from running
local CI with act with multiple jobs in parallel into separate logs.

For example,

contrib/devtools/run-local-ci.sh workflow=.github/workflows/cmake_production.yml job=depends-builds 2>&1 | tee cmake_production_depends_builds_parallel.log

would run all three production jobs in parallel.

Running contrib/devtools/split-logs.py cmake_production_depends_builds_parallel.log
afterwards would create three output files, each representing the
activity for a single job.

jamescowens committed:
Tweak ARM64 ARMhf CI jobs

Also --no-install-recommends for apt based OS's to prevent installing
unnecessary packages.

jamescowens committed:
New CI/CD documentation

jamescowens commented:
Getting pretty close on this. I need to straighten out the build documentation and then I will mark this ready for review.

jamescowens committed:
Add missing PARALLEL parameter help in build_targets.sh

jamescowens committed:
Enhance and align build.md documentation

jamescowens committed:
build: enable win64 cross-compile support for WSL in build scripts

This commit updates the helper scripts to support cross-compiling the
Windows target (win64) natively within the Windows Subsystem for Linux (WSL).

Changes to install_dependencies.sh:
- Added WSL detection logic (`check_is_wsl`).
- Conditional installation of `wine`:
  - Skipped on WSL (Windows host handles execution natively).
  - Installed on native Linux (required for testing).
- Fixed dependency logic to ensure `PKGS_BASE` (make, cmake, git, build-essential)
  is always installed for the `win64` target, as the depends system requires
  a full Linux toolchain.

Changes to build_targets.sh:
- Added logic to Step 4 (win64) to detect the execution environment.
- Dynamically sets `CMAKE_CROSSCOMPILING_EMULATOR`:
  - Sets to `/usr/bin/wine` on native Linux.
  - Unsets on WSL to utilize kernel-level binfmt_misc interoperability,
    allowing `.exe` files to run directly for tests.

jamescowens committed:
doc: add guide for building windows client on WSL

Adds `doc/build-windows-wsl.md`, a guide for building the Gridcoin
Windows client (64-bit) using the new CMake build scripts via WSL.

This documentation replaces the deprecated MSYS2 build instructions
and covers:
- Setting up the WSL environment.
- Using `build_targets.sh` with `TARGET=win64`.
- Creating the NSIS installer via the `deploy` target.
- Warnings regarding NSIS compressor failures on Debug builds.

jamescowens committed:
doc: add macOS build guide using Homebrew and CMake

This commit adds `doc/build-macos.md`, a comprehensive guide for building
the Gridcoin macOS client from source.

The guide reflects the current CI/CD environment configuration (`cmake_production.yml`)
and covers:
- Installing dependencies via Homebrew (Qt6, Boost, etc.).
- Configuring the build with CMake, dynamically detecting Homebrew prefixes.
- Compiling with explicit parallelism (`-j $(sysctl -n hw.logicalcpu)`).
- Running tests and creating the DMG installer using CPack.

Also removed deprecated build-unix.md and build-windows.md files.

jamescowens committed:
Change content of build-msys2.md file to indicate deprecation

jamescowens commented:
Ok. This is essentially done. Note that I may try to get OpenBSD and FreeBSD working but I won't hold up the review for that.

I ended up doing just about everything we need I think.

@CyberTailor, I ended up using a custom target for the NSIS installer (--target deploy) and mapping to the existing template. I did not want to mess with that file too much.

jamescowens committed:
build: add macOS support to unified build scripts

Updates `build_targets.sh` and `install_dependencies.sh` to fully support
macOS as a first-class build target (`TARGET=macos`).

Changes:
- install_dependencies.sh:
  - Added detection for macOS (`Darwin` kernel).
  - Implemented dependency installation via Homebrew.
  - Added logic to handle `USE_QT6` flag, installing either `qt` or `qt@5`.
  - Ensures correct packages are installed for macOS without sudo.

- build_targets.sh:
  - Added a new "Step 5" for macOS native builds.
  - Implemented dynamic detection of Homebrew prefixes for Qt (supporting both 5 and 6) and OpenSSL.
  - Added automatic detection of `icu4c` path to fix linker errors on macOS.
  - Added `EXTRA_CMAKE_ARGS` support to pass custom flags (e.g. static boost).
  - Configures CMake with correct paths and parallelism.
  - Updated help text and validation logic to include `macos` target.

jamescowens committed:
Update documentation on macOS build target

This updates the documentation to reflect the addition of the macos
target to build_targets.sh.

Also updates other documents to indicate current experimental or
deprecated status.

jamescowens commented:
I actually verified the macOS build and run in a Sonoma VM. Note that we need to fine tune the cmake package install (dmg) a bit.

jamescowens committed:
build: enhance install_dependencies.sh with GUI/Qt logic

Updates dependency installation logic to support optional GUI builds
and better handle Qt versioning.

Changes to install_dependencies.sh:
- Added `WITH_GUI` parameter to `install_deps` function.
- Conditional Qt Installation:
  - Only installs Qt packages if `WITH_GUI="true"`.
  - Supports `USE_QT6="false"` on macOS by installing `qt@5`.
  - Skips Qt installation on Linux if `USE_QT6="false"`, printing a warning
    that Qt5 dependencies must be managed manually (due to EOL status on
    many distros).
- Ensures `PKGS_BASE` (build-essential, cmake, etc.) is always installed.

- Minor tweaks to other files to adjust.

jamescowens committed:
Enhance split-logs.py script to satisfy Copilot review

jamescowens committed:
doc: rewrite OpenBSD build guide for CMake and Qt6

- Replace legacy build instructions with the modern CMake workflow.
- Add specific `pkg_add` dependency lists for both Headless and GUI builds.
- specific instructions for enabling Qt6 and XFCE integration.
- Include necessary system configuration steps (doas/sudo, wheel group, and messagebus).

jamescowens committed:
Add the ability to use Ubuntu 22.04 container in local ci

CyberTailor committed:
Build miniupnpc with CMake

CyberTailor committed:
toolchain.cmake tweaks

Changes summary:
1. Add WINDRES
2. Make sure OBJCOPY and OBJDUMP are set
3. Make sure DEPENDS_COMPILE_DEFINITIONS are actually used
4. Add library search path to default LDFLAGS
5. Always restrict pkg-config search path to depends
6. Remove irrelevant conditions

CyberTailor committed:
Restore gridcoinresearchd's Linux install steps

CyberTailor commented on CMakeLists.txt:
```diff
@@ -123,7 +135,7 @@ option(BUNDLED_QT5      "Use the bundled version of Qt 5" ${HUNTER_ENABLED})
 set(QT5_MINIMUM_VERSION 5.9.5)
 set(QT6_MINIMUM_VERSION 6.2.0)
-set(QT_COMPONENTS Core Concurrent Gui LinguistTools Network Widgets)
+set(QT_COMPONENTS Core Concurrent Gui LinguistTools Network Widgets Svg)
```
Is QtSvg a runtime dependency of Gridcoin?

CyberTailor committed:
Use CPack to generate NSIS installer

jamescowens commented:
Yes. Without it many icons will not appear in the depends builds.Sent from my iPhoneOn Dec 6, 2025, at 5:00 AM, Anna ***@***.***> wrote:﻿
@CyberTailor commented on this pull request.



In CMakeLists.txt:
> @@ -123,7 +135,7 @@ option(BUNDLED_QT5      "Use the bundled version of Qt 5" ${HUNTER_ENABLED})
 
 set(QT5_MINIMUM_VERSION 5.9.5)
 set(QT6_MINIMUM_VERSION 6.2.0)
-set(QT_COMPONENTS Core Concurrent Gui LinguistTools Network Widgets)
+set(QT_COMPONENTS Core Concurrent Gui LinguistTools Network Widgets Svg)

Is QtSvg a runtime dependency of Gridcoin?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

CyberTailor commented on CMakeLists.txt:
```diff
@@ -298,10 +352,18 @@ if(CMAKE_CXX_BYTE_ORDER EQUAL BIG_ENDIAN)
-check_include_file("sys/endian.h" HAVE_SYS_ENDIAN_H)
-check_include_file("sys/prctl.h" HAVE_SYS_PRCTL_H)
+if(CMAKE_CROSSCOMPILING AND WIN32)
+    message(STATUS "Cross-compiling for Windows: forcing POSIX headers to 'not found'")
```
Why is it necessary?

jamescowens replied:
Because on certain platforms with the cross-compile, the build was getting confused because it was finding the system headers instead. I would rather do this than hack the source code.

Many of these guards are necessary to be able to successfully compile on a "dirty" machine like a developer workstation that has all of the native development packages installed. This stuff often does not show up in a clean CI runner.

jamescowens commented:
I want to make a point to the reviewers. Many of the tweaks over the past week have been due to testing in different scenarios to ensure that the builds work across a wide variety of conditions. This is not limited to the CI checks, which themselves have posed a challenge. This includes testing on bare iron development machines with many libraries installed (good hermeticity checks), and VM's of various operating systems.

The CI checks themselves have posed challenges in terms of the correct dependencies. Additional complexity was generated by ensuring that the CI works in both act and Github actions, because the base runner images in Github are slightly different than the base containers used in act.

CyberTailor commented:
If I want to add my commits to this PR, should I sumbit each PR to the source branch individually or simply push new commits on top of a single PR?

jamescowens commented:
I have only a little bit more work to do on this unless you see something glaring. I am trying to get Ubuntu 22.04 to do the depends build for Windows (this is not for CI, since we use 24.04 in the runners). 22.04 has mingw 8, which has an incomplete set of headers for DirectX 12, and so is proving to be a big challenge with the Qt compile.

I also am pushing today updated documentation for OpenBSD and FreeBSD since I got both working using the native compile.

I am not going to do any more rebases or force pushes. The commits so far are the commits. Therefore you are good to base your work on the source branch for this PR.

I hope you agree this gives a good boost for us. I am sure you can see improvements to be made though.

jamescowens commented:
I don't think you have push permission to my branch on my fork, so you would have to do a PR to my fork at the moment.

We could also merge this and then you could do a fresh branch off of development. Quite frankly this should be merged, as it is much better than what we have currently and we need it for the next testnet builds to lead up to the Natasha release.

jamescowens committed:
Update OpenBSD and FreeBSD build guides based on testing with new build system

jamescowens committed:
Mark FreeBSD build instructions as stable but not checked in CI/CD

jamescowens committed:
Add xxd package to depends to ensure all hosts using version that accepts -n parameter

jamescowens committed:
Add warning guard for old mingw version for Ubuntu 22.04

Also add wine package to ensure tests run correctly.

jamescowens committed:
build: Fix depends fallback download logic to use sanitized filenames

This commit fixes a logic error in the package download fallback mechanism within
funcs.mk.

Previously, fallback attempts (e.g., to S3 or bitcoind-depends-sources) requested
the upstream filename/URL `$(3)`. This works when the upstream name matches the
local name, but fails for packages that rely on query strings or renaming
(e.g., `CMakeLists.txt?h=6.7.3`).

Mirrors are populated with the sanitized, locally-stored filenames `$(4)`. This change
forces fallback attempts to request `$(4)` instead of `$(3)`, ensuring the request matches
the storage convention on the mirrors and preventing 403/404 errors during primary download
failures.

jamescowens committed:
Merge pull request #48 from CyberTailor/depends-nsis-cpack

Use CPack to generate NSIS installer

jamescowens committed:
Add doas to linter ignore misspellings. (This is a valid OpenBSD command.)

jamescowens committed:
Remove trailing spaces on empty lines in install_dependencies.sh to satisfy linter.

jamescowens committed:
Add env.os_ref to the front of the depends cache key

This adds an OS_REF environment variable in the cmake_production.yml
file and then uses that at the front of the depends cache key. This
is to ensure that prior cached depends using a differnt runner OS
are invalidated rather than inappropriately reused.

jamescowens committed:
Fix copyright year variable and stripping of executables

jamescowens committed:
Add override to XXD path to ensure depends XXD is found during depends builds

This is the final cleanup for the Ubuntu 22.04 issue.

jamescowens committed:
Update build.md and build-windows-wsl.md

This updates build.md to make sure the path is specified for the
Windows installer command, and provides an Ubuntu 22.04 workaround
procedure for building the win64 in Ubuntu 22.04.
